### PR TITLE
feat(mock): added dats mock

### PR DESCRIPTION
**Pull request by antoniomuso**
## 🚨 Proposed changes

> Please review the [guidelines for contributing](../../CONTRIBUTING.md) to this repository.

Added Dats mock, now Dats can be tested without activating an UDP or TCP socket.
This feature can add breaking changes on types. 
## ⚙️ Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Bugfix (non-breaking change which fixes an issue)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation Update (if none of the other choices apply)
-   [ ] Refactor

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -29,25 +29,26 @@ export interface Options {
...
 /**
  * Statsd Client
  * @alias module:dats.Client
  */
 class Client {
-    private bufferSize: number;
-    private bufferFlushTimeout: number;
-    private socket: Socket;
-    private buffer: { length: number; data: string };
+    protected bufferSize: number;
```
[the remark this answers is not in the document]
Comment (antoniomuso) [reply src/index.ts:40]: To allow the extendability of Dats.

[the remark this answers is not in the document]
Comment (simonecorsi) [reply src/index.ts:40]: Can't we use a protected accessor only on the one that we need without modifying all the existing logic?

File: src/mock.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import Dats, { Options } from './index';
+import { Socket } from './socket';
+import { URL } from 'url';
+import { debuglog, DebugLoggerFunction } from 'util';
+
+export default class DatsMock extends Dats {
```
Comment (dnlup) [inline src/mock.ts:6]: Do we want to export this in  the whole module?

Comment (simonecorsi) [reply src/mock.ts:6]: I think this should be a separated export so we don't pollute the main class with `ifs`

Comment (dnlup) [reply src/mock.ts:6]: I think this needs just a little bit of documentation to explain how to access the mock.

**Review by dnlup — APPROVED** [review]

**Review by simonecorsi — CHANGES_REQUESTED** [review]
LGTM! But you have to write a section in the README on how to use it :)

File: src/mock.ts (hunk 2 above)
Comment (antoniomuso) [reply src/mock.ts:6]: Yes, I add it.

File: README.md [hunk 3]
```diff
@@ -201,9 +203,44 @@ All sending errors are handled by the `onError` callback.
...
+```ts
+import ClientMock from '@immobiliare/dats/mock';
+
+const host = new URL(`udp://127.0.0.1:8232`);
+const namespace = 'ns1';
+const client = new ClientMock({ host, namespace });
+
+client.gauge('some.metric', 100);
+client.set('some.metric', 100);
+// metrics is an array with all metrics send
```
Comment (simonecorsi) [inline README.md:219]: `sent`

File: README.md [hunk 4]
```diff
@@ -201,9 +203,44 @@ All sending errors are handled by the `onError` callback.
...
+console.log(client.metrics);
+/* stdout:
+    [
+        'ns1.some.metric:100|g',
+        'ns1.some.metric:100|s',
+    ]
+*/
+// Check if a metric is in the metrics array
+client.includes('ns1.some.metric:100|s'); // -> true
```
Comment (simonecorsi) [inline README.md:228]: Now that I see it, I'm thinking if something like `client.sent("abc")` is more correct 🤔

File: README.md [hunk 5]
```diff
@@ -201,9 +203,44 @@ All sending errors are handled by the `onError` callback.
...
+client.includes('ns1.some.metric:10|s'); // -> false
+client.includes(/ns1\.some\.metric:\d+\|s/); // -> true
+client.includes(/ns1\.some\.test:\d+\|s/); // -> false
+
+// Clean the metrics array with
+client.cleanMetrics();
```
Comment (simonecorsi) [inline README.md:234]: why not flush?

File: README.md (hunk 4 above)
Comment (antoniomuso) [reply README.md:228]: Mm, I like it. But my doubt is: do you think it is too similar to `send`?

Comment (simonecorsi) [reply README.md:228]: You got a point, we can use the verb like in most assertion libraries, something like `didSend`, `hasSend`. It'll make the test more readable!

**Comment by simonecorsi** [comment]
LGTM, nice job!

**Pull request merged**